File: backend/src/main/java/com/example/chat/controller/MessageController.java
package com.example.chat.controller;

import com.example.chat.entity.Message;
import com.example.chat.service.MessageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/messages")
public class MessageController {

    private final MessageService messageService;
    private final com.example.chat.service.ConversationService conversationService;
    private final com.example.chat.repository.UserRepository userRepository;
    private final com.example.chat.service.AiClient aiClient;

    @Autowired
    public MessageController(MessageService messageService,
                             com.example.chat.service.ConversationService conversationService,
                             com.example.chat.repository.UserRepository userRepository,
                             com.example.chat.service.AiClient aiClient) {
        this.messageService = messageService;
        this.conversationService = conversationService;
        this.userRepository = userRepository;
        this.aiClient = aiClient;
    }

    @PostMapping
    public ResponseEntity<?> sendMessage(@RequestBody MessageRequest request) {
        com.example.chat.entity.User user = userRepository.findById(request.getUserId()).orElse(null);
        if (user == null) {
            return ResponseEntity.badRequest().body("User not found with ID: " + request.getUserId());
        }

        com.example.chat.entity.Conversation conv = conversationService.getConversationById(request.getConversationId());
        if (conv == null) {
            return ResponseEntity.badRequest().body("Conversation not found with ID: " + request.getConversationId());
        }

        // 1. Save User Message
        Message userMessage = new Message();
        userMessage.setContent(request.getContent());
        userMessage.setUser(user);
        userMessage.setConversation(conv);
        messageService.saveMessage(userMessage);

        try {
            // 2. Call AI Service
            String aiResponseText = aiClient.sendMessage(user.getId().toString(), request.getContent());

            // 3. Save AI Message
            com.example.chat.entity.User aiUser = userRepository.findByUsername("ai_assistant");
            if (aiUser == null) {
                // Fallback or create if missing (though DataSeeder should handle it)
                aiUser = user; 
            }

            Message aiMessage = new Message();
            aiMessage.setContent(aiResponseText);
            aiMessage.setUser(aiUser);
            aiMessage.setConversation(conv);

            Message savedAiMessage = messageService.saveMessage(aiMessage);

            // Return the AI's message so the frontend can display it
            return ResponseEntity.ok(savedAiMessage);
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error processing AI response: " + e.getMessage());
        }
    }
    public static class MessageRequest {
        private String content;
        private Long conversationId;
        private Long userId;

        // Getters and Setters
        public String getContent() { return content; }
        public void setContent(String content) { this.content = content; }
        public Long getConversationId() { return conversationId; }
        public void setConversationId(Long conversationId) { this.conversationId = conversationId; }
        public Long getUserId() { return userId; }
        public void setUserId(Long userId) { this.userId = userId; }
    }

    @GetMapping("/{conversationId}")
    public ResponseEntity<List<Message>> getMessages(@PathVariable Long conversationId) {
        List<Message> messages = messageService.getMessagesByConversationId(conversationId);
        return ResponseEntity.ok(messages);
    }
}

File: backend/src/controllers/messageController.ts
import { Request, Response } from 'express';
import { messageService } from '../services/messageService';
import { conversationService } from '../services/conversationService';
import { authService } from '../services/authService';
import { aiClient } from '../services/aiClient';
import { CreateMessageRequest } from '../models/Message';

export class MessageController {
  public async sendMessage(req: Request, res: Response): Promise<void> {
    try {
      const { content, conversationId, userId }: CreateMessageRequest = req.body;

      if (!content) {
        res.status(400).send('Message content is required');
        return;
      }
      if (!conversationId) {
        res.status(400).send('Conversation ID is required');
        return;
      }
      if (!userId) {
        res.status(400).send('User ID is required');
        return;
      }

      const user = await authService.getUserById(userId);
      if (!user) {
        res.status(400).send(`User not found with ID: ${userId}`);
        return;
      }

      const conv = await conversationService.getConversationById(conversationId);
      if (!conv) {
        res.status(400).send(`Conversation not found with ID: ${conversationId}`);
        return;
      }

      // 1. Save User Message
      await messageService.saveMessage(conversationId, user.id, content);

      try {
        // 2. Call AI Service
        const aiResponseText = await aiClient.sendMessage(user.id.toString(), content);

        // 3. Save AI Message
        let aiUser = await authService.getUserByUsername('ai_assistant');
        if (!aiUser) {
          aiUser = user;
        }

        const savedAiMessage = await messageService.saveMessage(
          conversationId,
          aiUser.id,
          aiResponseText
        );

        // Return AI's message
        res.status(200).json(savedAiMessage);
      } catch (e: any) {
        res.status(500).send(`Error processing AI response: ${e.message}`);
      }
    } catch (error: any) {
      res.status(500).json({ error: error.message || 'Failed to send message' });
    }
  }

  public async getMessages(req: Request, res: Response): Promise<void> {
    try {
      const conversationId = parseInt(req.params.conversationId, 10);
      if (isNaN(conversationId)) {
        res.status(400).json({ error: 'Invalid conversation ID' });
        return;
      }

      const messages = await messageService.getMessagesByConversationId(conversationId);
      res.status(200).json(messages);
    } catch (error: any) {
      res.status(500).json({ error: error.message || 'Failed to retrieve messages' });
    }
  }
}

export const messageController = new MessageController();
